library/fluid/fluid_extensions/src/main/java/org/quiltmc/qsl/fluid/mixin/EntityMixin.java

```diff
@@ -0,0 +1,187 @@
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.vehicle.BoatEntity;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.tag.TagKey;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.random.RandomGenerator;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.World;
+import org.quiltmc.qsl.fluid.api.FlowableFluidExtensions;
+import org.quiltmc.qsl.fluid.impl.CustomFluidInteracting;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import javax.annotation.Nullable;
+
+
+@Mixin(Entity.class)
+public abstract class EntityMixin implements CustomFluidInteracting {
+	@Shadow
+	public float fallDistance;
+	@Shadow
+	public World world;
+	@Shadow
+	protected boolean firstUpdate;
+	@Shadow
+	@Final
+	protected RandomGenerator random;
+	protected boolean inCustomFluid;
+	protected boolean submergedInCustomFluid;
+	protected Fluid submergedCustomFluid;
+	@Shadow
+	private BlockPos blockPos;
+
+	@Shadow
+	public abstract boolean equals(Object o);
+
+	@Shadow
+	@Nullable
+	public abstract Entity getVehicle();
+
+	@Shadow
+	public abstract boolean updateMovementInFluid(TagKey<Fluid> tag, double d);
+
+	@Shadow
+	public abstract void extinguish();
+
+	@Shadow
+	public abstract double getEyeY();
+
+	@Shadow
+	public abstract double getX();
+
+	@Shadow
+	public abstract double getZ();
+
+	@Shadow
+	public abstract BlockPos getBlockPos();
+
+	@Shadow
+	public abstract boolean isSprinting();
+
+	@Shadow
+	public abstract boolean isSwimming();
+
+	@Shadow
+	public abstract void setSwimming(boolean swimming);
+
+	@Shadow
+	public abstract boolean hasVehicle();
+
+	@Shadow
+	public abstract double getY();
+
+	@Shadow
+	public abstract void setOnFireFromLava();
+
+	@Inject(method = "baseTick", at = @At("TAIL"))
+	public void baseTick(CallbackInfo ci) {
+		this.checkCustomFluidState();
+		this.updateSubmergedInCustomFluidState();
+	}
+
+	@Override
+	public boolean quilt$isInCustomFluid() {
+		return this.inCustomFluid;
+	}
+
+	@Override
+	public boolean quilt$isSubmergedInCustomFluid() {
+		return this.submergedInCustomFluid && this.quilt$isInCustomFluid();
+	}
+
+	void checkCustomFluidState() {
+		FluidState fluidState = this.world.getFluidState(this.getBlockPos());
+		this.submergedInCustomFluid = this.quilt$isSubmergedInCustomFluid(fluidState.getFluid());
+		if (fluidState.getFluid() instanceof FlowableFluidExtensions fluid &&
+				!(getVehicle() instanceof BoatEntity)) {
+			updateMovementInFluid(TagKey.of(Registry.FLUID_KEY,fluidState.getBuiltInRegistryHolder().getKey().get().getRegistry()), fluid.getPushStrength(fluidState, (Entity) (Object) this));
```

Comment:
add Javadoc to explain what the hell is going on here, please

Reply:
I'm actually not sure why I asked for Javadoc here, you can't really put Javadoc in the middle of a method...
Just turn these into normal block comments.